### PR TITLE
Adds the clang-format linter

### LINKED ISSUE #763: Add Support for clang-format
> "clang-format is located in clang/tools/clang-format and can be used to format C/C++/Java/JavaScript/JSON/Objective-C/Protobuf/C# code." ~ https://clang.llvm.org/docs/ClangFormat.html

Covering 6 different languages and 2 different serialization formats, this linter offers a lot of bang for your buck and seems like a good fit for Mega-Linter. It is one of a very small number of linters present in Super-Linter but missing from Mega-Linter.

The only officially supported installation method is [building from source](https://github.com/llvm/llvm-project/blob/main/clang/INSTALL.txt) in conjunction with all of LLVM, but unofficially it is also available as a [Docker image](https://hub.docker.com/r/unibeautify/clang-format/) and [Node.js package](https://github.com/Unibeautify/beautifier-clang-format) from [Unibeautify](https://unibeautify.com/).

### PR DESCRIPTION
**Pull request by daltonv**
Fixes #763

## Proposed Changes

1. Adds new clang-format linter

Updates the regular descriptor files for a new linter, but limits this linter to only be installed on the new c_cpp flavor, as the clang-extra-tools package is rather large. 

The c and cpp tests were updated to satisfy the clang-formats requirements too. 

### Author/Contributor
- [x] Add entry to the [CHANGELOG](https://github.com/oxsecurity/megalinter/blob/main/CHANGELOG.md) listing the change and linking to the corresponding issue (if appropriate)
- [x] If documentation is needed for this change, has that been included in this pull request

### Reviewing Maintainer
- [ ] Label as `breaking` if this is a large fundamental change
- [ ] Label as either `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`, or `performance`

### CODE REVIEW

**Comment by daltonv** [comment]
Not sure why the CI tests are failing. Looks like the r linter tests are the ones failing.

**Comment by daltonv** [comment]
Is there a way I can test this branch against another repo? ie run megalinter locally. 

I want to ensure the megalinter run picks up the `.clang-format` file in the root of a repo by default. I would expect it to do this, but it's always nice to test it for real.

**Comment by echoix** [comment]
> Is there a way I can test this branch against another repo? ie run megalinter locally. 
> 
> 
> 
> I want to ensure the megalinter run picks up the `.clang-format` file in the root of a repo by default. I would expect it to do this, but it's always nice to test it for real. 



If you build the docker image locally, you can then use `npx mega-linter-runner --nodockerpull --image <the name of your image:tag>` with other options to run it.

**Comment by echoix** [comment]
> Not sure why the CI tests are failing. Looks like the r linter tests are the ones failing. 

It was failing yesterday too https://github.com/oxsecurity/megalinter/pull/3086

**Comment by echoix** [comment]
Overall it looks really great!

I myself have other questions for @nvuillam :

- [ ] When does a linter descriptor need to include the way to call `--version`?

- [ ] Do we support having a linter descriptor copied two times in order to be in two files? Doesn't this create duplicate entries?

**Comment by daltonv** [comment]
> > Is there a way I can test this branch against another repo? ie run megalinter locally.
> > I want to ensure the megalinter run picks up the `.clang-format` file in the root of a repo by default. I would expect it to do this, but it's always nice to test it for real.
> 
> If you build the docker image locally, you can then use `npx mega-linter-runner --nodockerpull --image <the name of your image:tag>` with other options to run it.

Okay thanks for the tip. Good thing I tried it. There is actually an issue. 

I set `cli_config_arg_name: "--style"`. However to set a config file manually you must call `clang-format --style=file:.clang-format` note there cannot be a space between `--style=file:` and the file path. 

I updated my descriptor to use  ``cli_config_arg_name: "--style=file:"` however this does not fix it. When I run manually I see megalinter tries to run 
```
[clang-format] command: ['clang-format', '--Werror', '--dry-run', '--style=file:', '/tmp/lint/.clang-format', <my c files>
```

Is there a way I can describe in the descriptor to not add a space?

**Comment by daltonv** [comment]
Worst case scenario with the `cli_config_arg_name:` we could remove that and the `config_file_name` from the descriptor. `clang-format` itself will search for a root level `.clang-format` file.

File: megalinter/Linter.py [hunk 1]
```diff
@@ -1225,7 +1225,9 @@ def build_lint_command(self, file=None) -> list:
                 self.final_config_file = self.final_config_file.replace(
                     self.workspace, DEFAULT_DOCKER_WORKSPACE_DIR
                 )
-            if self.cli_config_arg_name.endswith("="):
+            if self.cli_config_arg_name.endswith(
+                "="
+            ) or self.cli_config_arg_name.endswith(":"):
```
Comment (daltonv) [inline megalinter/Linter.py:1230]: Okay I actually think I found a fix to the clang-format config arg issue here. Take a look and see if it makes sense.

I'm just not sure how to accurately test that this somehow doesn't break another linter.

**Comment by nvuillam** [comment]
>  When does a linter descriptor need to include the way to call --version?

When the argument is not `-v` :)

> Do we support having a linter descriptor copied two times in order to be in two files? Doesn't this create duplicate entries?

We current ML architecture we have no choice, we just have to make sure that the `install` block lines are exactly the same, to avoid duplicates in Dockerfile

File: megalinter/Linter.py (hunk 1 above)
Comment (nvuillam) [reply megalinter/Linter.py:1230]: CI jobs should tell you :)
It seems safe to me, as only descriptors with cli_config_arg_name ending by `:` should be impacted

Comment (nvuillam) [reply megalinter/Linter.py:1230]: In case it is too complicated, you can always define a python class that will override default methods :)

**Comment by daltonv** [comment]
Just rebased onto main again. Sorry for the force push. I'm used to the rebase workflow for syncing with main. Let me know and I can use merge commits in the future.

**Comment by nvuillam** [comment]
@daltonv we merge with "squash & merge" option so merge or rebase, the result is the same in the merge commit :)

**Comment by daltonv** [comment]
@nvuillam sounds good. Just wanted to ask in case it made it easier to review.

**Comment by nvuillam** [comment]
@daltonv I didn't want to merge just before a new release, but if you merge conflicts, I'll merge the PR and it will be available in beta, then in next release :)

**Comment by daltonv** [comment]
@nvuillam no worries. I'll fix the conflicts now.

**Comment by echoix** [comment]
Thanks a lot!!

**Comment by vkucera** [comment]
Thanks a lot @daltonv ! This is a significant improvement of MegaLinter's assets!

**Comment by vkucera** [comment]
Is the option to format files in place with the `-i` option implemented as well?

**Comment by daltonv** [comment]
@vkucera I didn't add the `-i` option. Taking another look at the descriptor schema though it should be simple for me to add that as well, I'll make a new issue to track this and I should be able to implement and test this before the next release.

**Comment by daltonv** [comment]
Adding `-i` support is tracked at #3137

**Comment by vkucera** [comment]
> @vkucera I didn't add the `-i` option. Taking another look at the descriptor schema though it should be simple for me to add that as well, I'll make a new issue to track this and I should be able to implement and test this before the next release.

Thanks a lot!

**Pull request merged**